R/Vietnam.R

```diff
@@ -0,0 +1,129 @@
+                         download = function() {
+                           bundles_urls = list(
+                             "case_by_time" = 'https://covid.ncsc.gov.vn/api/v3/covid/provinces?filter_type=case_by_time',
+                             "death_by_time" = 'https://covid.ncsc.gov.vn/api/v3/covid/provinces?filter_type=death_by_time',
+                             "recovered_by_time" = 'https://covid.ncsc.gov.vn/api/v3/covid/provinces?filter_type=recovered_by_time'
+                             
+                           ) 
+                           Sys.setenv("VROOM_CONNECTION_SIZE" = 131072*4) # Fix VROOM error
+                           provines_url = 'https://covid.ncsc.gov.vn/api/v3/covid/provinces'
+                           bundles = names(bundles_urls)
+                           provines_data = fromJSON(provines_url)
+                           
+                           get_bundles_data = function(bundles){
+                             bundles_data = list()
+                             for (bundle in bundles){
+                               url = paste0('https://covid.ncsc.gov.vn/api/v3/covid/provinces?filter_type=', bundle)
+                               data = fromJSON(url)
+                               bundles_data = c(bundles_data, setNames(list(data), bundle))
+                             }
+                             bundles_data
+                           }
+                           
+                           bundles_data = get_bundles_data(bundles)
+                           
+                           get_province = function(id, data){
+                             row_dat = provines_data[(id=id),]
+                             death_by_time= do.call(cbind, data$death_by_time[id])
+                             case_by_time=do.call(cbind, data$case_by_time[id])
+                             recovered_by_time=do.call(cbind, data$recovered_by_time[id])
+                             if (!identical(row.names(death_by_time), row.names(death_by_time))) {
+                               stop("Dates on case_by_time and death_by_time do not match!")
+                             }
+                             df = tibble(date= dmy(row.names(case_by_time)),
+                                         id = row_dat$id,
+                                         name = row_dat$name,
+                                         case_by_time= case_by_time,
+                                         death_by_time= death_by_time,
+                                         recovered_by_time= recovered_by_time)
+                             df
+                           }
+                           
+                           df = do.call(rbind, lapply(provines_data$id, function(id){get_province(id, bundles_data)}))
+                           names(df) <- c("date", "id", "region_name", "cases_total", "deaths_total", "recovered_total")
+                           self$data$raw[['main']] = df
```

Comment:
All of this feels like assembly of downloaded separate raw data files which would more naturally be done in the `clean_common` function.(Usually if you specify a list of CSV URLs, these are "magically" transformed into a list of tables in`self$data$raw` and then the `clean` functions will tidy them and rework them. `download` should download and do not much more.)